File: app/src/main/java/com/wenitech/cashdaily/ui/navigationView/NewClientFragment.java
package com.wenitech.cashdaily.ui.navigationView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;
import androidx.navigation.ui.NavigationUI;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Toast;

import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.google.firebase.firestore.FieldValue;
import com.wenitech.cashdaily.Data.model.Cliente;
import com.wenitech.cashdaily.R;
import com.wenitech.cashdaily.Util.StartedAddNewClient;
import com.wenitech.cashdaily.databinding.FragmentNewClientBinding;
import com.wenitech.cashdaily.viewModel.NewClientViewModel;

public class NewClientFragment extends Fragment {

    private FragmentNewClientBinding binding;
    private NewClientViewModel viewModel = new NewClientViewModel();

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentNewClientBinding.inflate(LayoutInflater.from(inflater.getContext()), container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull final View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewModel = new ViewModelProvider(this).get(NewClientViewModel.class);
        setupAdapterAutoCompleteTextViewGender();

        binding.buttonSaveClient.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                initRegisterNewClient();
            }
        });

        viewModel.getMessenge().observe(getViewLifecycleOwner(), new Observer<StartedAddNewClient>() {
            @Override
            public void onChanged(StartedAddNewClient startedAddNewClient) {
                switch (startedAddNewClient.getStatedAddNewClient()){
                    case StartedAddNewClient.STATED_ADD_NEW_CLIENT_INIT:

                        break;
                    case StartedAddNewClient.STATED_ADD_NEW_CLIENT_PROCESS:
                        showProgressBar(true);
                        break;
                    case StartedAddNewClient.STATED_ADD_NEW_CLIENT_SUCCESS:
                        //Navigation.findNavController(view).navigateUp();
                        showProgressBar(false);
                        onSucces();
                        break;
                    case StartedAddNewClient.STATED_ADD_NEW_CLIENT_FAILED:
                        showProgressBar(false);
                        onError();
                        break;
                    case StartedAddNewClient.STATED_ADD_NEW_CLIENT_CANCEL:
                        showProgressBar(false);
                        break;
                }
            }
        });
    }

    private void setupAdapterAutoCompleteTextViewGender() {
        String[] opcionesGenero = new String[]{"Mujer", "Hombre", "No espesificar", "Otro"};
        ArrayAdapter<String> adapter = new ArrayAdapter<>(getContext(), R.layout.item_dropdown, opcionesGenero);
        binding.autoCompleteTextViewGender.setAdapter(adapter);
    }


    public void initRegisterNewClient() {
        if (isFormValid()) {
            String fullName = binding.editTextFullName.getText().toString().trim();
            String idClient = binding.editTextId.getText().toString().trim();
            String gender = binding.autoCompleteTextViewGender.getText().toString().trim();
            String phoneNumber = binding.editTextPhoneNumber.getText().toString().trim();
            String city = binding.editTextCity.getText().toString().trim();
            String direction = binding.editTextDirection.getText().toString().trim();

            // Todo:Enviar Objeto Cliente al viewModel
            viewModel.addNewClient(fullName,idClient,gender,phoneNumber,city,direction);
        }
    }

    /**
     * Check if each editText is filled correctly
     * if not insert an error message.
     *
     * @return true if ok or false if not
     */
    private boolean isFormValid() {
        boolean valid = true;

        if (TextUtils.isEmpty(binding.editTextFullName.getText())) {
            binding.editTextFullName.setError("Escribe un nombre");
            valid = false;
        } else if (TextUtils.isEmpty(binding.editTextId.getText())) {
            binding.editTextId.setError("Identificacion del cliente");
            valid = false;
        } else if (TextUtils.isEmpty(binding.autoCompleteTextViewGender.getText())) {
            binding.autoCompleteTextViewGender.setError("Elige una opcion");
            valid = false;
        } else if (TextUtils.isEmpty(binding.editTextPhoneNumber.getText())) {
            binding.editTextPhoneNumber.setError("Escribe un numero de contacto");
            valid = false;
        } else if (TextUtils.isEmpty(binding.editTextCity.getText())) {
            binding.editTextCity.setError("Agrega un ciudad");
            valid = false;
        } else if (TextUtils.isEmpty(binding.editTextDirection.getText())) {
            binding.editTextDirection.setError("Direcion de residencia");
            valid = false;
        }
        return valid;
    }

    private void showProgressBar(Boolean showProgress) {
        if (showProgress) {
            binding.groupContent.setVisibility(View.INVISIBLE);
            binding.groupProgresbar.setVisibility(View.VISIBLE);
        } else {
            binding.groupContent.setVisibility(View.VISIBLE);
            binding.groupProgresbar.setVisibility(View.INVISIBLE);
        }
    }

    private void onSucces() {
        Snackbar.make(binding.getRoot(),"Cliente agregado", BaseTransientBottomBar.LENGTH_SHORT).show();
    }

    private void onError() {
        Snackbar.make(binding.getRoot(),"Ocurio un error", BaseTransientBottomBar.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/com/wenitech/cashdaily/framework/features/client/newClient/NewClientFragment.kt
package com.wenitech.cashdaily.framework.features.client.newClient

import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.navigation.fragment.findNavController
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.Snackbar
import com.wenitech.cashdaily.R
import com.wenitech.cashdaily.domain.common.Resource
import com.wenitech.cashdaily.databinding.FragmentNewClientBinding
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class NewClientFragment : Fragment() {

    private val viewModel by viewModels<NewClientViewModel>()

    private lateinit var _binding: FragmentNewClientBinding
    private val binding get() = _binding

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        _binding = FragmentNewClientBinding.inflate(
            LayoutInflater.from(inflater.context),
            container,
            false
        )
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setupAdapterAutoCompleteTextViewGender()
        binding.buttonSaveClient.setOnClickListener { initRegisterNewClient() }

        observerLiveData()
    }

    private fun observerLiveData() {

        viewModel.resourceSaveClient.observe(viewLifecycleOwner, Observer {
            when (it) {
                is com.wenitech.cashdaily.domain.common.Resource.Failure -> {
                    showProgressBar(false)
                    onError(it.msg.toString())
                }
                is com.wenitech.cashdaily.domain.common.Resource.Loading -> {
                    showProgressBar(true)
                }
                is com.wenitech.cashdaily.domain.common.Resource.Success -> {
                    showProgressBar(false)
                    onSucces(it.data.toString())
                    findNavController().navigateUp()
                }
            }
        })
    }

    private fun setupAdapterAutoCompleteTextViewGender() {
        val opcionesGenero = arrayOf("Mujer", "Hombre", "No espesificar", "Otro")
        val adapter = ArrayAdapter(requireContext(), R.layout.item_dropdown, opcionesGenero)
        binding.autoCompleteTextViewGender.setAdapter(adapter)
    }

    fun initRegisterNewClient() {
        if (isFormValid) {
            val fullName = binding.editTextFullName.text.toString().trim { it <= ' ' }
            val cedulaClient = binding.editTextId.text.toString().trim { it <= ' ' }
            val gender = binding.autoCompleteTextViewGender.text.toString().trim { it <= ' ' }
            val phoneNumber = binding.editTextPhoneNumber.text.toString().trim { it <= ' ' }
            val city = binding.editTextCity.text.toString().trim { it <= ' ' }
            val direction = binding.editTextDirection.text.toString().trim { it <= ' ' }

            // Todo:Enviar Objeto Cliente al viewModel
            viewModel.saveNewClient(fullName, cedulaClient, gender, phoneNumber, city, direction)
        }
    }

    /**
     * Check if each editText is filled correctly
     * if not insert an error message.
     *
     * @return true if ok or false if not
     */
    private val isFormValid: Boolean
        private get() {
            var valid = true
            if (TextUtils.isEmpty(binding.editTextFullName.text)) {
                binding.editTextFullName.error = "Escribe un nombre"
                valid = false
            } else if (TextUtils.isEmpty(binding.editTextId.text)) {
                binding.editTextId.error = "Identificacion del cliente"
                valid = false
            } else if (TextUtils.isEmpty(binding.autoCompleteTextViewGender.text)) {
                binding.autoCompleteTextViewGender.error = "Elige una opcion"
                valid = false
            } else if (TextUtils.isEmpty(binding.editTextPhoneNumber.text)) {
                binding.editTextPhoneNumber.error = "Escribe un numero de contacto"
                valid = false
            } else if (TextUtils.isEmpty(binding.editTextCity.text)) {
                binding.editTextCity.error = "Agrega un ciudad"
                valid = false
            } else if (TextUtils.isEmpty(binding.editTextDirection.text)) {
                binding.editTextDirection.error = "Direcion de residencia"
                valid = false
            }
            return valid
        }

    private fun showProgressBar(showProgress: Boolean) {
        if (showProgress) {
            binding.groupContent.visibility = View.INVISIBLE
            binding.groupProgresbar.visibility = View.VISIBLE
        } else {
            binding.groupContent.visibility = View.VISIBLE
            binding.groupProgresbar.visibility = View.INVISIBLE
        }
    }

    private fun onSucces(dato: String) {
        Snackbar.make(binding.root, dato, BaseTransientBottomBar.LENGTH_LONG).show()
    }

    private fun onError(dato: String) {
        Snackbar.make(binding.root, dato, BaseTransientBottomBar.LENGTH_SHORT).show()
    }
}